### PR TITLE
goroutine: Make monitoring background routines context-aware

### PR DESCRIPTION
**Pull request by efritz**
Add a context parameter to the `MonitorBackgroundRoutines` function. Cancelling the context will have the same effect as a single Ctrl+C. Two Ctrl+C's in a row will still trigger an immediate application exit.

### CODE REVIEW

File: internal/goroutine/background.go [hunk 1]
```diff
@@ -19,32 +20,33 @@ type StartableRoutine interface {
...
 	var wg sync.WaitGroup
 	startAll(&wg, routines...)
-	waitForSignal(signals)
+	waitForSignal(ctx, signals)
 	stopAll(&wg, routines...)
 	wg.Wait()
 }
 
-// startAll calls each routine's Start method in its own goroutine and and registers
+// startAll calls each routine's Start method in its own goroutine and and registersq
```
Comment (eseliger) [inline internal/goroutine/background.go:49]:
```suggestion
// startAll calls each routine's Start method in its own goroutine and registers
```

File: internal/goroutine/background.go [hunk 2]
```diff
@@ -64,21 +66,31 @@ func stopAll(wg *sync.WaitGroup, routines ...BackgroundRoutine) {
...
+// waitForSignal blocks until the given context is canceled or signal has been
+// received on the given channel. If two signals are received, os.Exit(0) will
+// be called immediately.
+func waitForSignal(ctx context.Context, signals <-chan os.Signal) {
+	select {
+	case <-ctx.Done():
+		go exitOnSignal(signals, 2)
+
+	case <-signals:
+		go exitOnSignal(signals, 1)
```
Comment (eseliger) [inline internal/goroutine/background.go:78]: I read this as
> If the context is canceled, try proceeding once more
> If we receive a signal, cancel immediately

Shouldn't it be the other way around? 🤔

File: internal/goroutine/background.go [hunk 3]
```diff
@@ -64,21 +66,31 @@ func stopAll(wg *sync.WaitGroup, routines ...BackgroundRoutine) {
...
-		// Shutdown immediately on a second signal
+// exitOnSignal waits for n signals on the given channel, then calls os.Exit(0).
+func exitOnSignal(signals <-chan os.Signal, n int) {
+	for i := 0; i < n; i++ {
 		<-signals
-		exiter()
-	}()
+	}
+
+	exiter()
```
Comment (eseliger) [inline internal/goroutine/background.go:93]: Oh, I didn't know it does that. Probably this should not be used in a goroutine in the campaigns background file right? (What we discussed recently)

File: internal/goroutine/background.go (hunk 2 above)
Comment (efritz) [reply internal/goroutine/background.go:78]: Ah so if context is canceled then you wait for 2 signals then exit. if you get a signal you wait for 1 more signal then exit.

Would a better name for that function clear it up?

Comment (eseliger) [reply internal/goroutine/background.go:78]: Ahh. `ExitAfterSignals` maybe? Maybe its just me though

**Review by eseliger — APPROVED** [review]
rebase error?

**Pull request merged**